### DISCUSSION TITLE
Trigger `on_reaction_add` on messages from history.

### DISCUSSION THREAD

**Discussion by JackTemaki** [Rapptz/discord.py#8369] [Q&A] [answered]
I am writing a small bot that should display if a user added a reaction to messages. The problem is that `on_reaction_add` is only triggered for messages that are written after the bot was activated. I tried to load a (limited) history of messages on startup with:
```python
class ExampleClient(discord.Client):
    async def on_ready(self):
         for channel in discord.Client.get_all_c:hannels(self):
            channel = await self.fetch_channel(channel.id)
            print(f"Channel {channel.name} with ID  {channel.id}")
            if hasattr(channel, "history"):
                async for message in channel.history(limit=10):
                    message = await channel.fetch_message(message.id)
                    print(f"cached {message.id}")
```
Which does show me the messages I want to react on. Still, no `on_reaction_add` is triggered. 
I saw some JavaScript tutorials that message caching should help, so what am I missing here?

Simple test case of my `on_reaction_add`:

```python
    async def on_reaction_add(self, reaction: discord.Reaction, user):
        print(f'User {user} added reaction {reaction} in channel {reaction.message.channel}')
        bot_channel = discord.Client.get_channel(self, 1011016620321296425)
        await bot_channel.send(content=f"A rating of {reaction} was placed in {reaction.message.channel} for link {reaction.message.content}")
```

**Comment by i-am-zaidali** [comment, marked as answer]
`on_reaction_add` is only triggered by messae that are available in cache of the bot. You should instead be leveraging the [`on_raw_reaction_add`](https://discordpy.readthedocs.io/en/stable/api.html?highlight=on_raw_reaction_add#discord.on_raw_reaction_add) event which gets triggered regardless of a message's cache presence. However this event does not take a reaction, user pair as argument but instead takes a payload that includes appropriate information for you to handle the event as you like

**Reply by JackTemaki** [reply]
Thanks a lot! Was not aware of this function, works perfectly!

Just for completeness, would there be a way to manually fill the cache like I intended to do?

**Reply by LostLuma** [reply]
Adding items to the cache is not a supported operation at this time.
